includes/class-indieauth-authenticate.php

```diff
@@ -81,50 +100,78 @@ public function verify_access_token( $token ) {
 		return $params['me'];
 	}
 
+	public static function generate_state() {
+		$state = wp_generate_password( 128, false );
+		$value = wp_hash( $state, 'nonce' );
+		setcookie( 'indieauth_state', $value, current_time( 'timestamp' ) + 120, '/', false, true );
+		return $state;
+	}
+
 	/**
 	 * Redirect to Authorization Endpoint for Authentication
 	 *
 	 * @param string $me URL parameter
 	 * @param string $redirect_uri where to redirect
 	 *
 	 */
-	public static function authorization_redirect( $me, $redirect_uri ) {
+	public function authorization_redirect( $me, $redirect_uri ) {
+		$endpoints = indieauth_discover_endpoint( $me );
+		if ( ! $endpoints ) {
+			return new WP_Error(
+				'authentication_failed',
+				__( '<strong>ERROR</strong>: Could not discover endpoints', 'indieauth' ),
+				array(
+					'status' => 401,
+				)
+			);
+		}
+		$authorization_endpoint = null;
+		if ( isset( $endpoints['authorization_endpoint'] ) ) {
+			$authorization_endpoint = $endpoints['authorization_endpoint'];
+		}
+		$state = $this->generate_state();
 		$query = build_query(
 			array(
-				'me'           => rawurlencode( $me ),
-				'redirect_uri' => wp_login_url( $redirect_uri ),
-				'client_id'    => home_url(),
-				'state'        => wp_create_nonce( 'indieauth-' . home_url() ),
+				'me'            => rawurlencode( $me ),
+				'redirect_uri'  => rawurlencode( $redirect_uri ),
+				'client_id'     => rawurlencode( home_url() ),
+				'state'         => $state,
+				'response_type' => 'id',
 			)
 		);
 		// redirect to authentication endpoint
-		wp_redirect( get_option( 'indieauth_authorization_endpoint' ) . '?' . $query );
+		wp_redirect( $authorization_endpoint . '?' . $query );
```

Comment:
you should use a more solid method to add the query to the authorization endpoint. If hoster does not support `mod_rewrite`, `$authorization_endpoint` might already has query params.